File: supabase/functions/_shared/academusoft_pdf.ts
/**
 * Academusoft 4.0 «Listado de Estudiantes Inscritos» — ADR-0002
 * Mirror of src/services/academusoft_pdf.py for Supabase Edge (Deno).
 */
import { extractText, getDocumentProxy } from "npm:unpdf@0.12.1";

const FOOTER_MARKERS = [
  "señor docente",
  "legalizar matrícula",
  "legalizar matricula",
];

const ROW_RE =
  /^\s*(\d+)\s+(?:TI|CC)\s*-\s*(\d+)\s+\S+\s+(.+?)\s*$/i;
const MATERIA_LINE_RE = /^(.+?)\s+([0-9][0-9A-Z_]+)\s*$/i;
const COURSE_CODE_RE = /^([A-Z0-9]+)-/i;

export type AcademusoftPdfHeader = {
  materia: string;
  course_code: string;
  group_name: string;
};

export type AcademusoftPdfRow = {
  roster_position: number;
  document_number: string;
  full_name: string;
};

export type AcademusoftPdfParseResult = {
  header: AcademusoftPdfHeader;
  rows: AcademusoftPdfRow[];
};

export class AcademusoftPdfError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AcademusoftPdfError";
  }
}

export class AcademusoftPdfModuleMismatchError extends AcademusoftPdfError {
  constructor(
    public pdfMateria: string,
    public pdfGroup: string,
    public moduleCourseCode: string,
    public moduleGroup: string,
  ) {
    super(
      `Este PDF es de ${pdfMateria} / ${pdfGroup}; usted está en ${moduleCourseCode} / ${moduleGroup}.`,
    );
    this.name = "AcademusoftPdfModuleMismatchError";
  }
}

function normalizeToken(value: string): string {
  return value
    .trim()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/\s+/g, " ")
    .toUpperCase();
}

async function extractPdfText(content: Uint8Array): Promise<string> {
  const pdf = await getDocumentProxy(content);
  const { text } = await extractText(pdf, { mergePages: true });
  if (!text || !text.trim()) {
    throw new AcademusoftPdfError("PDF vacío o sin texto extraíble");
  }
  return text;
}

function isFooterLine(line: string): boolean {
  const lowered = line.trim().toLowerCase();
  return FOOTER_MARKERS.some((marker) => lowered.includes(marker));
}

function parseHeader(lines: string[]): AcademusoftPdfHeader {
  for (let index = 0; index < lines.length; index++) {
    const line = lines[index];
    if (!line.toLowerCase().includes("materia") || !line.toLowerCase().includes("grupo")) {
      continue;
    }
    const materiaLine = (lines[index + 1] ?? "").trim();
    const match = MATERIA_LINE_RE.exec(materiaLine);
    if (!match) {
      throw new AcademusoftPdfError(
        `No se pudo leer Materia/Grupo del PDF: ${materiaLine}`,
      );
    }
    const materia = match[1].trim();
    const groupName = match[2].trim();
    const codeMatch = COURSE_CODE_RE.exec(materia);
    if (!codeMatch) {
      throw new AcademusoftPdfError(`No se pudo leer el código de materia: ${materia}`);
    }
    return {
      materia,
      course_code: codeMatch[1].toUpperCase(),
      group_name: groupName,
    };
  }
  throw new AcademusoftPdfError("Encabezado Materia/Grupo no encontrado en el PDF");
}

function parseRowsFromText(text: string): AcademusoftPdfRow[] {
  const rows: AcademusoftPdfRow[] = [];
  for (const line of text.split(/\r?\n/)) {
    if (isFooterLine(line)) break;
    const match = ROW_RE.exec(line);
    if (!match) continue;
    rows.push({
      roster_position: Number(match[1]),
      document_number: match[2],
      full_name: match[3].trim(),
    });
  }
  return rows;
}

export async function parseAcademusoftPdf(
  content: Uint8Array,
): Promise<AcademusoftPdfParseResult> {
  const text = await extractPdfText(content);
  const lines = text.split(/\r?\n/);
  const header = parseHeader(lines);
  const rows = parseRowsFromText(text);

  if (!rows.length) {
    throw new AcademusoftPdfError("No se encontraron estudiantes en el PDF");
  }

  const seen = new Set<number>();
  const deduped: AcademusoftPdfRow[] = [];
  for (const row of [...rows].sort((a, b) => a.roster_position - b.roster_position)) {
    if (seen.has(row.roster_position)) continue;
    seen.add(row.roster_position);
    deduped.push(row);
  }

  return { header, rows: deduped };
}

export function validatePdfHeaderForModule(
  header: AcademusoftPdfHeader,
  courseCode: string,
  groupName: string,
): void {
  if (normalizeToken(header.group_name) !== normalizeToken(groupName)) {
    throw new AcademusoftPdfModuleMismatchError(
      header.materia,
      header.group_name,
      courseCode,
      groupName,
    );
  }
  const materiaNorm = normalizeToken(header.materia);
  const codeNorm = normalizeToken(courseCode);
  if (!materiaNorm.startsWith(`${codeNorm}-`)) {
    if (normalizeToken(header.course_code) !== codeNorm) {
      throw new AcademusoftPdfModuleMismatchError(
        header.materia,
        header.group_name,
        courseCode,
        groupName,
      );
    }
  }
}

File: src/services/academusoft_pdf.py
"""Parse Academusoft 4.0 «Listado de Estudiantes Inscritos» PDF reports (ADR-0002)."""

from __future__ import annotations

import io
import re
import unicodedata
from dataclasses import dataclass

import pdfplumber

_FOOTER_MARKERS = (
    "señor docente",
    "legalizar matrícula",
    "legalizar matricula",
)
_HEADER_ROW_LABELS = ("no.", "documento", "código", "codigo", "nombre completo")
_ROW_RE = re.compile(
    r"^\s*(\d+)\s+(?:TI|CC)\s*-\s*(\d+)\s+\S+\s+(.+?)\s*$",
    re.IGNORECASE,
)
_MATERIA_LINE_RE = re.compile(
    r"^(.+?)\s+([0-9][0-9A-Z_]+)\s*$",
    re.IGNORECASE,
)
_COURSE_CODE_RE = re.compile(r"^([A-Z0-9]+)-", re.IGNORECASE)


@dataclass(frozen=True)
class AcademusoftPdfHeader:
    materia: str
    course_code: str
    group_name: str


@dataclass(frozen=True)
class AcademusoftPdfRow:
    roster_position: int
    document_number: str
    full_name: str


@dataclass(frozen=True)
class AcademusoftPdfParseResult:
    header: AcademusoftPdfHeader
    rows: list[AcademusoftPdfRow]


class AcademusoftPdfError(ValueError):
    """Base error for PDF parsing or validation."""


class AcademusoftPdfModuleMismatchError(AcademusoftPdfError):
    """PDF Materia/Grupo do not match the open module."""

    def __init__(self, pdf_materia: str, pdf_group: str, module_course_code: str, module_group: str):
        self.pdf_materia = pdf_materia
        self.pdf_group = pdf_group
        self.module_course_code = module_course_code
        self.module_group = module_group
        super().__init__(
            "Este PDF es de "
            f"{pdf_materia} / {pdf_group}; "
            f"usted está en {module_course_code} / {module_group}."
        )


def _normalize_token(value: str) -> str:
    folded = unicodedata.normalize("NFKD", value.strip())
    ascii_only = "".join(ch for ch in folded if not unicodedata.combining(ch))
    return re.sub(r"\s+", " ", ascii_only).upper()


def _extract_text(content: bytes) -> str:
    with pdfplumber.open(io.BytesIO(content)) as pdf:
        if not pdf.pages:
            raise AcademusoftPdfError("PDF vacío")
        return "\n".join(page.extract_text() or "" for page in pdf.pages)


def _is_footer_line(line: str) -> bool:
    lowered = line.strip().lower()
    return any(marker in lowered for marker in _FOOTER_MARKERS)


def _parse_header(lines: list[str]) -> AcademusoftPdfHeader:
    for index, line in enumerate(lines):
        if "materia" in line.lower() and "grupo" in line.lower():
            if index + 1 >= len(lines):
                break
            materia_line = lines[index + 1].strip()
            match = _MATERIA_LINE_RE.match(materia_line)
            if not match:
                raise AcademusoftPdfError(
                    f"No se pudo leer Materia/Grupo del PDF: {materia_line!r}"
                )
            materia = match.group(1).strip()
            group_name = match.group(2).strip()
            code_match = _COURSE_CODE_RE.match(materia)
            if not code_match:
                raise AcademusoftPdfError(f"No se pudo leer el código de materia: {materia!r}")
            return AcademusoftPdfHeader(
                materia=materia,
                course_code=code_match.group(1).upper(),
                group_name=group_name,
            )
    raise AcademusoftPdfError("Encabezado Materia/Grupo no encontrado en el PDF")


def _parse_rows_from_tables(content: bytes) -> list[AcademusoftPdfRow]:
    rows: list[AcademusoftPdfRow] = []
    with pdfplumber.open(io.BytesIO(content)) as pdf:
        for page in pdf.pages:
            for table in page.extract_tables() or []:
                if not table:
                    continue
                header_idx = None
                for idx, raw_row in enumerate(table):
                    cells = [str(cell or "").strip().lower() for cell in raw_row]
                    if "documento" in cells and any("nombre" in cell for cell in cells):
                        header_idx = idx
                        break
                if header_idx is None:
                    continue
                for raw_row in table[header_idx + 1 :]:
                    if not raw_row or len(raw_row) < 4:
                        continue
                    line = " ".join(str(cell or "").strip() for cell in raw_row if cell)
                    parsed = _ROW_RE.match(line)
                    if parsed:
                        rows.append(
                            AcademusoftPdfRow(
                                roster_position=int(parsed.group(1)),
                                document_number=parsed.group(2),
                                full_name=parsed.group(3).strip(),
                            )
                        )
    return rows


def _parse_rows_from_text(text: str) -> list[AcademusoftPdfRow]:
    rows: list[AcademusoftPdfRow] = []
    for line in text.splitlines():
        if _is_footer_line(line):
            break
        match = _ROW_RE.match(line)
        if not match:
            continue
        rows.append(
            AcademusoftPdfRow(
                roster_position=int(match.group(1)),
                document_number=match.group(2),
                full_name=match.group(3).strip(),
            )
        )
    return rows


def parse_academusoft_pdf(content: bytes) -> AcademusoftPdfParseResult:
    text = _extract_text(content)
    lines = text.splitlines()
    header = _parse_header(lines)

    rows = _parse_rows_from_tables(content)
    if len(rows) < 1:
        rows = _parse_rows_from_text(text)

    if not rows:
        raise AcademusoftPdfError("No se encontraron estudiantes en el PDF")

    seen_positions: set[int] = set()
    deduped: list[AcademusoftPdfRow] = []
    for row in sorted(rows, key=lambda item: item.roster_position):
        if row.roster_position in seen_positions:
            continue
        seen_positions.add(row.roster_position)
        deduped.append(row)

    return AcademusoftPdfParseResult(header=header, rows=deduped)


def validate_pdf_header_for_module(
    header: AcademusoftPdfHeader,
    *,
    course_code: str,
    group_name: str,
) -> None:
    if _normalize_token(header.group_name) != _normalize_token(group_name):
        raise AcademusoftPdfModuleMismatchError(
            header.materia,
            header.group_name,
            course_code,
            group_name,
        )
    if not _normalize_token(header.materia).startswith(_normalize_token(course_code) + "-"):
        if _normalize_token(header.course_code) != _normalize_token(course_code):
            raise AcademusoftPdfModuleMismatchError(
                header.materia,
                header.group_name,
                course_code,
                group_name,
            )
